---
title: "Analyse des données avec Phyloseq"
author : Mélody Lebrun
date: "3 janvier 2022"
output: 
  github_document:
    toc: true
    toc_depth: 2
---


# Préparation de l'environnement

```{r}
library(phyloseq)
library(ggplot2)
library(readr)

```

```{r}
sapply(c(.cran_packages, .bioc_packages), require, character.only = TRUE)
```

```{r}
load("02_Data-analysis-with-DADA2_FinalEnv")
```

# Construire un arbre phylogénétique permettant de relié les variants de séquences

```{r}
seqs <- getSequences(seqtabNoC)
names(seqs) <- seqs 
alignment <- AlignSeqs(DNAStringSet(seqs), anchor=NA,verbose=FALSE)
```

```{r}
phangAlign <- phyDat(as(alignment, "matrix"), type="DNA")
dm <- dist.ml(phangAlign)
treeNJ <- NJ(dm)
fit = pml(treeNJ, data=phangAlign)
fitGTR <- update(fit, k=4, inv=0.2)
fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE,
        rearrangement = "stochastic", control = pml.control(trace = 0))
detach("package:phangorn", unload=TRUE)
```

# Combiner des données dans un objet phyloseq

```{r}
Metadonnees<-read.csv(file="SraRunTable(2).txt")
```

```{r}
Metadonnees$Sample.Name <- paste0(gsub("00", "", Metadonnees$Sample.Name), "D")
Metadonnees <- Metadonnees[!duplicated(Metadonnees$Sample.Name),] # Remove dupicate entries for reverse reads
rownames(seqtabAll) <- gsub("124", "125", rownames(seqtabAll)) # Fix discrepancy
all(rownames(seqtabAll) %in% Metadonnees$Sample.Name) # TRUE
```

```{r}
rownames(Metadonnees) <- Metadonnees$Sample.Name
keep.cols <- c("Run","env_medium","host_issue_sampled","lat_lon") 
```